<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [53]:
import pandas as pd
import psycopg2

In [106]:
DBNAME = None
USER = None
PASSWORD = None
HOST = None
PORT = None

In [55]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [56]:
# сохраняем в переменую запрос
# запрос через select
# обращаемся к таблице с помщью from
query_3_1 = f'''select count(name)
            from vacancies
            '''

In [57]:
# переменной присваиваем запрос и выводим его
df_3_1 = pd.read_sql_query(query_3_1, connection);
df_3_1

C:\Users\vivec\AppData\Local\Temp\ipykernel_10332\1820195450.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_1 = pd.read_sql_query(query_3_1, connection);


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [58]:
query_3_2 = f'''select count(name)
                from employers
                '''

In [59]:
df_3_2 = pd.read_sql_query(query_3_2, connection);
df_3_2

C:\Users\vivec\AppData\Local\Temp\ipykernel_10332\2467592708.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_2 = pd.read_sql_query(query_3_2, connection);


count
0  23501

3. Посчитайте с помощью запроса количество регионов (таблица areas).

In [60]:
query_3_3 = f'''select count(name)
                from areas
                '''

In [61]:
df_3_3 = pd.read_sql_query(query_3_3, connection);
df_3_3

C:\Users\vivec\AppData\Local\Temp\ipykernel_10332\65068916.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_3 = pd.read_sql_query(query_3_3, connection);


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [62]:
query_3_4 = f'''select count(name)
                from industries
                '''

In [63]:
df_3_4 = pd.read_sql_query(query_3_4, connection);
df_3_4

C:\Users\vivec\AppData\Local\Temp\ipykernel_10332\1084753995.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_4 = pd.read_sql_query(query_3_4, connection);


count
0    294

***

### 1. Получили общее количество вакансий представленных в таблице (49197)
### 2. Получили количество работадателей (23501)
### 3. Получили количество регионов (1362)
### 4. Получили количество сфер деятельностей в базе (294)


# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [64]:
# соединяем таблицы по уникальным значениям id с помощью join 
# делаем группировку по региону с помощью group by
# сортируем запрос по количеству в порядке убывания вакансий с помощью order by
# задайм лимит с помощью limit 
query_4_1 = f'''select 
                    a.name as area,
                    count(v.id) as cnt
                from areas as a
                    join vacancies as v on a.id = v.area_id
                group by area
                order by cnt desc
                limit 5
                '''

In [65]:
df_4_1 = pd.read_sql_query(query_4_1, connection);
df_4_1

C:\Users\vivec\AppData\Local\Temp\ipykernel_10332\2632720463.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_1 = pd.read_sql_query(query_4_1, connection);


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [66]:
# задаём условия по зарплате с помощью where
query_4_2 = f'''select 
                    count(v.id) as cnt
                from vacancies as v
                where v.salary_from is not null 
                    or v.salary_to is not null
                '''

In [67]:
df_4_2 = pd.read_sql_query(query_4_2, connection);
df_4_2

C:\Users\vivec\AppData\Local\Temp\ipykernel_10332\3369437776.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_2 = pd.read_sql_query(query_4_2, connection);


cnt
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [68]:
# получаем среднее значение порогов ЗП и округляем его
query_4_3 = f'''select 
                    round(avg(v.salary_from)) as asf,
                    round(avg(v.salary_to)) as ast
                from vacancies as v
                '''

In [69]:
df_4_3 = pd.read_sql_query(query_4_3, connection);
round(df_4_3)

C:\Users\vivec\AppData\Local\Temp\ipykernel_10332\1578565216.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_3 = pd.read_sql_query(query_4_3, connection);


asf       ast
0  71065.0  110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [70]:
query_4_4 = f'''select
                    v.schedule,
                    v.employment,
                    count(*)
                from vacancies as v
                group by v.schedule, v.employment
                order by 3 desc
                '''

In [71]:
df_4_4 = pd.read_sql_query(query_4_4, connection);
df_4_4

C:\Users\vivec\AppData\Local\Temp\ipykernel_10332\1830085806.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_4 = pd.read_sql_query(query_4_4, connection);


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [72]:
query_4_5 = f'''select
                    v.experience,
                    count(v.id)
                from vacancies as v
                group by v.experience
                having v.experience is not null
                order by 2
                '''

In [73]:
df_4_5 = pd.read_sql_query(query_4_5, connection);
df_4_5

C:\Users\vivec\AppData\Local\Temp\ipykernel_10332\1727186623.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_5 = pd.read_sql_query(query_4_5, connection);


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

### 1. Мы узнали в каких городах больше всего вакансий (Москва, Санкт-Петербург и Минск)
### 2. Выяснили, что у 24073 вакансии указан хотя бы один порог ЗП.
### 3. Узнали среднее значение по минимальной и макимальной вилке ЗП.
### 4. Составили таблицу с данными по количетсву вакансий в сочетании график работы/ тип занятости (больше всего вакансий Полный день/ Полная занятость)
### 5. Получили данные с требованием к опыту работы и количеству вакансий, большая часть вакансий с требованием от 1 до 3 лет, меньшая часть более 6 лет опыта работы.
#### По этим данным видно, что большое количество вакансий в столицах Росии и Беларусси, почти у половни вакансий указан порог ЗП, большая часть вакансий приходится на полный день/ полная занятость, более 50% вакансий требуют опыт работы от 1 до 3 лет, но так же имеется и много вакансий без Опыта работы, значит работадатели готовы принимать специалистов, которые только отучились или ещё учатся.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [74]:
query_5_1 = f'''select
                    e.name,
                    count(v.id)
                from employers as e
                    join vacancies as v on v.employer_id = e.id
                group by e.name
                order by 2 desc
                limit 5
                '''

In [75]:
df_5_1 = pd.read_sql_query(query_5_1, connection);
df_5_1

C:\Users\vivec\AppData\Local\Temp\ipykernel_10332\3591465723.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_1 = pd.read_sql_query(query_5_1, connection);


name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [76]:
# после группировки задаём условия с помощью having
query_5_2 = f'''select 
                    a.name as reg_name,
                    count(e.id) as cnt_employer,
                    count(v.id) as cnt_vacancies
                from areas as a
                    full join vacancies as v on v.area_id = a.id
                    full join employers as e on e.area = a.id
                group by 1
                having count(v.id) = 0
                order by 2 desc
                '''

In [77]:
df_5_2 = pd.read_sql_query(query_5_2, connection);
df_5_2

C:\Users\vivec\AppData\Local\Temp\ipykernel_10332\3464236946.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_2 = pd.read_sql_query(query_5_2, connection);


reg_name  cnt_employer  cnt_vacancies
0                          Россия           410              0
1                       Казахстан           207              0
2              Московская область            75              0
3              Краснодарский край            19              0
4                        Беларусь            18              0
..                            ...           ...            ...
588                      Струнино             0              0
589    Славгород (Алтайский край)             0              0
590                     Сорочинск             0              0
591  Ленинский (Тульская область)             0              0
592                         Тавда             0              0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [78]:
# с помощью distinct получаем уникальные значения 
query_5_3 = f'''select
                    e.name as emp_name,
                    count(distinct v.area_id) as cnt_area
                from vacancies as v
                    join employers as e on e.id = v.employer_id
                group by e.id
                order by 2 desc
                '''

In [79]:
df_5_3 = pd.read_sql_query(query_5_3, connection);
df_5_3

C:\Users\vivec\AppData\Local\Temp\ipykernel_10332\3136109519.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_3 = pd.read_sql_query(query_5_3, connection);


emp_name  cnt_area
0                                      Яндекс       181
1                                  Ростелеком       152
2                                  Спецремонт       116
3                      Поляков Денис Иванович        88
4                                   ООО ЕФИН         71
...                                       ...       ...
14901                            НПП Авиатрон         1
14902              Центр дистанционных торгов         1
14903  Городские Телекоммуникационные Системы         1
14904                       Введенский, Отель         1
14905              Группа Компаний Автоимпорт         1

[14906 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [80]:
query_5_4 = f'''select
                    count(e.id)
                from employers as e
                    full join employers_industries as ei on ei.employer_id = e.id
                where ei.industry_id is null
                '''

In [81]:
df_5_4 = pd.read_sql_query(query_5_4, connection);
df_5_4

C:\Users\vivec\AppData\Local\Temp\ipykernel_10332\980203943.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_4 = pd.read_sql_query(query_5_4, connection);


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [82]:
query_5_5 = f'''select
                    e.name as comp_name,
                    count(ei.industry_id) as cnt_indusrty
                from employers as e
                    full join employers_industries as ei on ei.employer_id = e.id
                group by comp_name
                having count(ei.industry_id) = 4
                order by 1
                '''

In [83]:
df_5_5 = pd.read_sql_query(query_5_5, connection);
df_5_5

C:\Users\vivec\AppData\Local\Temp\ipykernel_10332\4268956696.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_5 = pd.read_sql_query(query_5_5, connection);


comp_name  cnt_indusrty
0        101 Интернет             4
1            21vek.by             4
2                2ГИС             4
3                  2К             4
4         4 пикселя +             4
...               ...           ...
1131            ЮРИОН             4
1132  ЮТИП Технологии             4
1133           ЯКласс             4
1134            ЯрНео             4
1135    Яшин&Партнёры             4

[1136 rows x 2 columns]

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [84]:
query_5_6 = f'''select
                    count(distinct e.id)
                from employers as e
                    full join employers_industries as ei on ei.employer_id = e.id
                    full join industries as i on ei.industry_id = i.id
                where i.name = 'Разработка программного обеспечения'
                '''

In [85]:
df_5_6 = pd.read_sql_query(query_5_6, connection);
df_5_6

C:\Users\vivec\AppData\Local\Temp\ipykernel_10332\1956844454.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_6 = pd.read_sql_query(query_5_6, connection);


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [86]:
# код для получения списка городов-милионников
import requests
from bs4 import BeautifulSoup

# переменной присваиваем адресс страницы
url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров'
# читаем адресс страницы 
response = requests.get(url) 
page = BeautifulSoup(response.text, 'html.parser')
# получаем таблицы 
list_city = page.find_all('table') 
# выбираем нужную таблицу
list_city = list_city[1] 
# создаём пустой список для городов
cities_ml = []
# перебираем все значения и заносим в список
for i in list_city.find_all('td'): 
    city = i.text
    cities_ml.append(city)
# отбираем только города из полученного списка 
cities_ml = cities_ml[1::5] 
# заменяем значение 
cities_ml[13] = 'Воронеж' 
cities_ml = list(cities_ml)
print(cities_ml)

['Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград']


In [87]:
# делаем объединение таблиц с помощью union all, чтобы добавить строку Total
query_5_7 = f'''select
                    a.name as name_reg,
                    count(distinct v.id) as cnt_vac
                from areas as a
                    join vacancies as v on v.area_id = a.id
                    join employers as e on v.employer_id = e.id 
                    and e.name = 'Яндекс'
                    and a.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 
                    'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск', 
                    'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 
                    'Краснодар', 'Воронеж', 'Пермь', 'Волгоград')
                group by 1
                union all
                select 
                    'Total',
                    count(*)
                from areas as a
                    join vacancies as v on v.area_id = a.id
                    join employers as e on v.employer_id = e.id 
                    and e.name = 'Яндекс'
                    and a.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 
                    'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск', 
                    'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 
                    'Краснодар', 'Воронеж', 'Пермь', 'Волгоград')
                '''

In [88]:
df_5_7 = pd.read_sql_query(query_5_7, connection);
df_5_7

C:\Users\vivec\AppData\Local\Temp\ipykernel_10332\3244254936.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_7 = pd.read_sql_query(query_5_7, connection);


name_reg  cnt_vac
0             Total      485
1         Волгоград       24
2           Воронеж       32
3      Екатеринбург       39
4            Казань       25
5         Краснодар       30
6        Красноярск       23
7            Москва       54
8   Нижний Новгород       36
9       Новосибирск       35
10             Омск       21
11            Пермь       25
12   Ростов-на-Дону       25
13           Самара       26
14  Санкт-Петербург       42
15              Уфа       26
16        Челябинск       22

***

### 1. Получили таблицу с указаннием работадателей у которых больше всего вакансий, на первом месте Яндекс, за ним следует Ростелеком и Тинькофф.
### 2. Получили таблицу с регионами у которых много работадателей, но нет вакансий, возможно в регионе и не требуются специалисты, а возможно и была допущена ошибка в вакансии.
### 3. Получили список работадателей и количество регионов в которых есть вакансии, на первом месте Яндекс, на втором Ростелеком и на третьем Спецремонт.
### 4. Узнали что у 8419 работадателей не указана сфера деятельности.
### 5. Получили таблицу с работадателями у которых указанно 4 сфер деятельности (1136 работадателей)
### 6. Узнали что 3553 работадателя ищут разработчика программного обеспечения.
### 7. Собрали данные по компании Яндекс, выяснили сколько вакансий в регеонах миллионниках и получили общее количество вакансий (485)
### Из этих данных можно сделать вывод, что компания Яндекс предоставляет большое количество вакансий в разных регионах, компания стремительно развивается.  

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [89]:
# в условиях поиска названия приводим к нижнему регистру с помощью lower
# делаем поиск по совпадению слов в названии вакансии 
query_6_1 = f'''select
                    count(v.id)
                from vacancies as v 
                where lower(v.name) like '%данн%'
                    or lower(v.name) like '%data%'
                '''
                

In [90]:
df_6_1 = pd.read_sql_query(query_6_1, connection);
df_6_1

C:\Users\vivec\AppData\Local\Temp\ipykernel_10332\1490021324.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_1 = pd.read_sql_query(query_6_1, connection);


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [91]:
query_6_2 = f'''select
                    count(v.name)
                from vacancies v
                where (lower(v.name) like '%data%scientist%'
                    or lower(v.name) like '%data%science%'
                    or lower(v.name) like '%исследователь%данных%'
                    or (v.name like '%ML%'
                    and v.name not like '%HTML%')
                    or lower(v.name) like '%machine%learning%'
                    or lower(v.name) like '%машинн%обучен%')
                    and (lower(v.experience) like '%нет%опыта%' 
                    or lower(v.employment) like '%стажировка%'
                    or lower(v.name) like '%junior%')
                '''

In [92]:
df_6_2 = pd.read_sql_query(query_6_2, connection);
df_6_2

C:\Users\vivec\AppData\Local\Temp\ipykernel_10332\683851843.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_2 = pd.read_sql_query(query_6_2, connection);


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [93]:
query_6_3 = f'''select
                    count(v.name)
                from vacancies v
                where (lower(v.name) like '%data%scientist%'
                or lower(v.name) like '%data%science%'
                or lower(v.name) like '%исследователь%данных%'
                or (v.name like '%ML%'
                    and v.name not like '%HTML%')
                or lower(v.name) like '%machine%learning%'
                or lower(v.name) like '%машинн%обучен%')
                and (v.key_skills like '%SQL%'
                    or lower(v.key_skills) like '%postgres%')
                '''

In [94]:
df_6_3 = pd.read_sql_query(query_6_3, connection);
df_6_3

C:\Users\vivec\AppData\Local\Temp\ipykernel_10332\3746522555.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_3 = pd.read_sql_query(query_6_3, connection);


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [95]:
query_6_4 = f'''select
                    count(v.name)
                from vacancies v
                where (lower(v.name) like '%data%scientist%'
                or lower(v.name) like '%data%science%'
                or lower(v.name) like '%исследователь%данных%'
                or (v.name like '%ML%'
                    and v.name not like '%HTML%')
                or lower(v.name) like '%machine%learning%'
                or lower(v.name) like '%машинн%обучен%')
                and lower(v.key_skills) like '%python%'
                '''

In [96]:
df_6_4 = pd.read_sql_query(query_6_4, connection);
df_6_4

C:\Users\vivec\AppData\Local\Temp\ipykernel_10332\2873102221.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_4 = pd.read_sql_query(query_6_4, connection);


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [97]:
# получаем среднее значение с помощью avg
# так же получаем длину строки с помощью length 
# замения символы в строке с помощью replace 
query_6_5 = f'''select
                    avg(
                    length(v.key_skills)
                    - length(replace(v.key_skills, CHR(9), ''))+1
                    )
                from vacancies v
                where (lower(v.name) like '%data%scientist%'
                    or lower(v.name) like '%data%science%'
                    or lower(v.name) like '%исследователь%данных%'
                    or (v.name like '%ML%'
                        and v.name not like '%HTML%')
                    or lower(v.name) like '%machine%learning%'
                    or lower(v.name) like '%машинн%обучен%')
                    and v.key_skills is not null
                '''

In [98]:
df_6_5 = pd.read_sql_query(query_6_5, connection);
round(df_6_5,2)

C:\Users\vivec\AppData\Local\Temp\ipykernel_10332\2433854141.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_5 = pd.read_sql_query(query_6_5, connection);


avg
0  6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [99]:
# используем coalesce, если есть значение null, вернет int
query_6_6 = f'''select
                    distinct(v.experience),
                    avg(coalesce((v.salary_to + v.salary_from)/2,
                    (v.salary_from), (v.salary_to)))
                from vacancies v
                where (lower(v.name) like '%data%scientist%'
                or lower(v.name) like '%data%science%'
                or lower(v.name) like '%исследователь%данных%'
                or (v.name like '%ML%'
                    and v.name not like '%HTML%')
                or lower(v.name) like '%machine%learning%'
                or lower(v.name) like '%машинн%обучен%')
                and (v.salary_to is not null
                    or v.salary_from is not null)
                group by 1
                '''

In [100]:
df_6_6 = pd.read_sql_query(query_6_6, connection);
round(df_6_6)

C:\Users\vivec\AppData\Local\Temp\ipykernel_10332\2417512424.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_6 = pd.read_sql_query(query_6_6, connection);


experience       avg
0           Нет опыта   74643.0
1  От 1 года до 3 лет  139675.0
2       От 3 до 6 лет  243115.0

***

### 1. Выяснили, 1771 вакансий связаны с данными.
### 2. Выяснили, есть 51 вакансий для junior специалиста в области Data Scientist.
### 3. Выяснили, в 201 вакансии по DS ищут специалистов с основными навыками SQL или postgres.
### 4. Выяснили, в 351 вакансии по DS ищут специалистов с основным навыком Pyrhon.
### 5. Выяснили среднее значение (6.41) по количеству указанных навыков по DS.
### 6. Выяснили среднюю ЗП по вакансиям DS от опыта работы (Нет опыт: 74643, От 1 до 3 лет: 139675, От 3 до 6 лет: 243115)

# Общий вывод по проекту

In [101]:
add_query = f'''select
                    distinct(a.name) as area_name,
                    e.name as employer_name,
                    v.name vacancie_name,
                    avg(
                        coalesce(
                            (v.salary_from + v.salary_to)/2,
                            (v.salary_from),
                            (v.salary_to)
                            )
                            ) as avg_salary,
                    count(v.name) as cnt_vac
                from vacancies as v
                    left join areas as a on a.id = v.area_id
                    left join employers as e on e.id = v.employer_id
                group by 1, 2, 3, v.salary_from, v.salary_to
                having (lower(v.name) like '%data%scientist%'
                or lower(v.name) like '%data%science%'
                or lower(v.name) like '%исследователь%данных%'
                or (v.name like '%ML%'
                    and v.name not like '%HTML%')
                or lower(v.name) like '%machine%learning%'
                or lower(v.name) like '%машинн%обучен%')
                and (v.salary_from is not null
                    or v.salary_to is not null)
                order by 1, avg_salary desc
                '''
df_add = pd.read_sql_query(add_query, connection);
df_add

C:\Users\vivec\AppData\Local\Temp\ipykernel_10332\1070365431.py:28: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_add = pd.read_sql_query(add_query, connection);


area_name                                      employer_name  \
0            Алматы                                       Kolesa Group   
1            Алматы                                Cybernet (Кибернет)   
2           Армения                                           Andersen   
3           Армения                                           Andersen   
4           Барнаул                                            Энтерра   
5          Белгород                                 Зеленая долина, ГК   
6       Владивосток                                  DNS Головной офис   
7       Владивосток                                  DNS Головной офис   
8      Екатеринбург                                          Сима-ленд   
9      Екатеринбург                         Сталепромышленная компания   
10     Екатеринбург  КБ Уральский банк реконструкции и развития (УБ...   
11           Казань                                              Ранкс   
12             Кипр                                            Semrush   
13          Королев                                        РКК Энергия   
14           Москва                                               Итон   
15           Москва                                         AppQuantum   
16           Москва                                          JSA Group   
17           Москва                                         СберЛизинг   
18           Москва                                             Dbrain   
19           Москва                                          КИБЕР-РОМ   
20           Москва                                       Кловер Групп   
21           Москва                        Freedom International Group   
22           Москва                                       Spice Agency   
23           Москва                                             Bip.ru   
24           Москва                                         АЙТИ.СПЕЙС   
25           Москва                                             Телега   
26           Москва                                         DIGINETICA   
27           Москва                                         JuicyScore   
28           Москва                            Алгоритмические Системы   
29           Москва                                       УЦ Новый шаг   
30           Москва            Центр хранения и анализа больших данных   
31           Москва                           Интеллектуальные решения   
32           Москва                                       НСК ЭкоСфера   
33           Москва                                     РТК Автоматика   
34           Москва                                  Азимут Технолоджи   
35           Москва                                        РКК Энергия   
36           Москва                                            Лидстех   
37           Москва                              ВижнЛабс (VisionLabs)   
38           Москва                                  Cpaaggregator.com   
39           Москва                                  Строительный Двор   
40      Новосибирск                                               Дром   
41      Новосибирск                           Igor Polyakov Consulting   
42      Новосибирск                                            Апэрбот   
43            Пермь                                           Homework   
44            Пермь                                            Випакс+   
45   Ростов-на-Дону                                          Theyr Ltd   
46           Рязань                                         MNOGOGAMES   
47           Рязань                                 Программный регион   
48           Рязань                                 Программный регион   
49  Санкт-Петербург                                          JSA Group   
50  Санкт-Петербург                                            OVISION   
51  Санкт-Петербург                                       Hype Auditor   
52  Санкт-Петербург                                              НИИАС   
53  Санкт-Пете

In [102]:
add_query_1 = (
                f'''select
                    a.name as area_name,
                    v.name as vac_name,
                    count(v.id)
                from vacancies as v 
                    join areas as a on a.id = v.area_id
                group by 1, 2
                having lower(v.name) like '%данн%'
                    or lower(v.name) like '%data%'
                order by 3 desc, 1
                '''
                )
df_add_1 = pd.read_sql_query(add_query_1, connection);
df_add_1

C:\Users\vivec\AppData\Local\Temp\ipykernel_10332\1888952356.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_add_1 = pd.read_sql_query(add_query_1, connection);


area_name                          vac_name  count
0              Москва                   Аналитик данных     41
1              Москва                    Data Scientist     28
2              Москва                     Data Engineer     18
3              Москва                      Data Analyst     14
4     Санкт-Петербург                      Data Analyst     11
...               ...                               ...    ...
1422             Ялта      Инженер сети передачи данных      1
1423        Ярославль      Аналитик данных (sql, excel)      1
1424        Ярославль           Ведущий аналитик данных      1
1425        Ярославль    Инженер группы передачи данных      1
1426        Ярославль  Старший администратор баз данных      1

[1427 rows x 3 columns]

In [103]:
add_query_2 = (
                f'''select
                    distinct(v.name),
                    count(v.name)
                    from vacancies as v
                    group by 1
                    order by 2 desc
                    limit 50
                    '''
                )
df_add_2 = pd.read_sql_query(add_query_2, connection);
df_add_2

C:\Users\vivec\AppData\Local\Temp\ipykernel_10332\777482712.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_add_2 = pd.read_sql_query(add_query_2, connection);


name  count
0                             Системный администратор   1761
1                                      Программист 1С   1153
2                                     Бизнес-аналитик    488
3                    Специалист технической поддержки    457
4                                  Системный аналитик    432
5                                            Аналитик    404
6                                 Инженер-программист    389
7                                      Программист 1C    360
8                                Frontend-разработчик    242
9                       Инженер технической поддержки    230
10                                        Программист    224
11                               Руководитель проекта    191
12            Младший специалист ручного тестирования    169
13        Специалист службы поддержки в Яндекс Бизнес    168
14                                  Менеджер проектов    153
15                                   Java-разработчик    150
16          Специалист по информационной безопасности    148
17  Специалист службы поддержки Рекламной сети Янд...    139
18                                    PHP-разработчик    135
19          Специалист службы поддержки пользователей    126
20                                Компьютерный мастер    123
21                                   Менеджер проекта    120
22                               Технический писатель    111
23                                Маркетолог-аналитик    108
24                              Руководитель проектов    106
25                             Ведущий программист 1C    106
26                             Технический специалист    104
27         Технический писатель (дистанционный автор)     99
28                 Сотрудник поддержки поиска Яндекса     99
29                    Ведущий системный администратор     99
30                Специалист поддержки в Yandex Cloud     97
31                             Ведущий программист 1С     94
32                               Технический директор     94
33                        Ведущий инженер-программист     92
34                                        QA Engineer     91
35                                    Web-разработчик     88
36                 Помощник системного администратора     86
37                                    IOS разработчик     86
38                                        Аналитик 1С     85
39                            Программист-разработчик     85
40  Специалист службы поддержки с техническими зна...     84
41                                     Разработчик 1С     83
42                                Компьютерный Мастер     83
43                                     Консультант 1С     82
44                                    Web-программист     81
45                                    Project Manager     81
46                       Ученик компьютерного мастера     79
47                                    Аналитик данных     78
48                                    Сетевой инженер     78
49                                            Инженер     78

In [104]:
add_query_3 = (
                f'''select 
                        i.name,
                        count(v.id)
                    from vacancies as v
                        join employers_industries as ei on v.employer_id = ei.employer_id
                        join industries as i on i.id = ei.industry_id
                        group by 1
                        order by 2 desc
                        limit 10
                        '''
                        )
df_add_3 = pd.read_sql_query(add_query_3, connection);
df_add_3

C:\Users\vivec\AppData\Local\Temp\ipykernel_10332\3335318323.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_add_3 = pd.read_sql_query(add_query_3, connection);


name  count
0                Разработка программного обеспечения  12499
1  Системная интеграция,  автоматизации технологи...  11034
2  Интернет-компания (поисковики, платежные систе...   6413
3                                               Банк   2742
4                                    Мобильная связь   1585
5                                 Кадровые агентства   1143
6                              Консалтинговые услуги   1137
7                                 Интернет-провайдер   1095
8                       Розничная сеть (продуктовая)   1034
9                                Фиксированная связь   1028

### В дополнительном иследованнии получил данные о регионе/ работадателе/ вакансии DS/ среднюю ЗП/ количеству вакансий, при условии, что хотя бы один порог ЗП был указан, а так же получил таблицу по поиску специалистов в области данных и количества вакансий по специальностям. Как видно по таблице большее количество вакансий в России, самую крупную ЗП предлагают в МСК и составляет она 450.000. Как видно вакансий в области данных 1771, вакансий в области DS 483, полностью заполненых вакансий DS 61 , что говорит о потребности специалистов в этой области, но и так же говорит о большой потребности и в других областях IT. Составил таблицу рейтинг по количеству разных вакансий, большую часть топа заняли именно сферы IT. Такая большая потребность специалистов говорит о том, что IT технологии по России развиваются и вакансий будет всё больше, из-за чего возможно повысят минимальной и максимальный порог ЗП.

In [105]:
# не забываем закрыть соединение после окончания работы
connection.close()